services/gitlab/gitlab-coverage.service.js

```diff
@@ -0,0 +1,149 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const { coveragePercentage } = require('../color-formatters')
+const { optionalUrl } = require('../validators')
+const { BaseSvgScrapingService, NotFound } = require('..')
+
+const schema = Joi.object({
+  message: Joi.string()
+    .regex(/^([0-9]+\.[0-9]+%)|unknown$/)
+    .required(),
+}).required()
+
+const queryParamSchema = Joi.object({
+  gitlab_url: optionalUrl,
+  job_name: Joi.string(),
+}).required()
+
+const documentation = `
+<p>
+  Important: If your project is publicly visible, but the badge is like this:
+  <img src="https://img.shields.io/badge/coverage-not&nbsp;set&nbsp;up-red" alt="coverage not set up"/>
+</p>
+<p>
+  Check if your pipelines are publicly visible as well.<br />
+  Navigate to your project settings on GitLab and choose General Pipelines under CI/CD.<br />
+  Then tick the setting Public pipelines.
+</p>
+<p>
+  Now your settings should look like this:
+</p>
+<img src="https://user-images.githubusercontent.com/12065866/67156911-e225a180-f324-11e9-93ad-10aafbb3e69e.png" alt="Setting Public pipelines set"/>
+<p>
+Also make sure you have set up code covrage parsing as described <a href="https://docs.gitlab.com/ee/ci/pipelines/settings.html#test-coverage-parsing">here</a>
+</p>
+<p>
+  Your badge should be working fine now.
+</p>
+`
+
+module.exports = class GitlabCoverage extends BaseSvgScrapingService {
+  static get category() {
+    return 'coverage'
+  }
+
+  static get route() {
+    return {
+      base: 'gitlab/coverage',
+      pattern: ':user/:repo/:branch',
+      queryParamSchema,
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Gitlab code coverage',
+        namedParams: {
+          user: 'gitlab-org',
+          repo: 'gitlab-runner',
+          branch: 'master',
+        },
+        staticPreview: this.render({ coverage: 67 }),
+        documentation,
+      },
+      {
+        title: 'Gitlab code coverage (specific job)',
+        namedParams: {
+          user: 'gitlab-org',
+          repo: 'gitlab',
+          branch: 'master',
+        },
+        queryParams: { job_name: 'rspec:coverage' },
```

Comment:
Out of curiosity, do we know of any jobs that have this setup? 

The upstream badge for this is returning `uknown` so the Shields badge is showing the corresponding `not set up`
https://gitlab.com/gitlab-org/gitlab/badges/master/coverage.svg?job=rspec:coverage
![](https://gitlab.com/gitlab-org/gitlab/badges/master/coverage.svg?job=rspec:coverage)

Reply:
This job did indeed exist at least at the time of writing (https://gitlab.com/gitlab-org/gitlab/-/jobs/667537958), but it's possible that it was recently removed

Reply:
We could use the gitlab-runner repo instead, that has lesser activity

https://gitlab.com/gitlab-org/gitlab-runner/-/jobs/670879955

Reply:
While we strive to have examples that work/fully demonstrate their respective features, I was primarily interested in finding a working instance of a coverage badge for a specific job. Would `test coverage report` be the correct job name for the gitlab-runner` repo? 

This one does indeed seem to work:
https://shields-staging-pr-5262.herokuapp.com/gitlab/coverage/gitlab-org/gitlab-runner/master?job_name=test%20coverage%20report
![](https://shields-staging-pr-5262.herokuapp.com/gitlab/coverage/gitlab-org/gitlab-runner/master?job_name=test%20coverage%20report)

Reply:
Okay, I'll use that. While I would've liked to have the example show a job in case where different jobs had different coverages (`gitlab-org/gitlab` had different coverage for Ruby and JS), I think we can give the higher priority to a working badge instead.

Reply:
Yeah that would've been ideal, but for now just having something that works will suffice :smile: